In [1]:
"""
Script Name: transform_HUB_data.py
Authores: Walter Arredondo, Carlos Daza, Andrés Coral
Fecha: 2026-10-03
Fase ETL: Transformación
Descripción:
Extrae los datos de Capacidad, generación y consumo de electricidad de HUB de Energia,
https://hubenergia.org/index.php/es/indicators/capacidad-generacion-y-consumo-de-electricidad,
guardado en data/raw/ realiza las transformaciones necesarias y las guarda como archivo CSV para posterior carga.
Entradas: - CSV con los resultados de la extracción: data/raw/
extract_HUb_generacion_electrica.csv
extract_HUb_capacidad_instalada.csv
extract_HUb_consumo.csv
Salidas: - Archivo CSV con las transformaciones realizadas: data/transformed/HUB_energia.csv
Notas: - Ejecutar este script una sola vez
"""

'\nScript Name: transform_HUB_data.py\nAuthores: Walter Arredondo, Carlos Daza, Andrés Coral\nFecha: 2026-10-03\nFase ETL: Transformación\nDescripción:\nExtrae los datos de Capacidad, generación y consumo de electricidad de HUB de Energia,\nhttps://hubenergia.org/index.php/es/indicators/capacidad-generacion-y-consumo-de-electricidad,\nguardado en data/raw/ realiza las transformaciones necesarias y las guarda como archivo CSV para posterior carga.\nEntradas: - CSV con los resultados de la extracción: data/raw/\nextract_HUb_generacion_electrica.csv\nextract_HUb_capacidad_instalada.csv\nextract_HUb_consumo.csv\nSalidas: - Archivo CSV con las transformaciones realizadas: data/processed/HUB_energia.csv\nNotas: - Ejecutar este script una sola vez\n'

# Recuperar información de la extracción
A continuación se cargan los csv generados por la fase de extracción.:


1.   HUB_generacion_electrica.csv
2.   HUB_capacidad_instalada.csv
3.   HUB_consumo.csv



In [2]:
import pandas as pd

In [3]:
from google.colab import drive

# Montar Google Drive
drive.mount('/content/drive')

Mounted at /content/drive


# HUB_generacion_electrica.csv

In [4]:
ruta = "/content/drive/My Drive/Colab Notebooks/Master/1.ETL/entregable3/data/raw/extract_HUB_generacion_electrica.csv"
df1_consolidado = pd.read_csv(ruta, sep = ';')

In [5]:
# Verificar la cantidad de registros y el número de columnas del dataframe df_consolidado
df1_consolidado.shape

(3960, 4)

Al final se obtienen 3960 registros con 4columnas, la misma información recuperada en la extracción de datos.

In [6]:
# Conocer nombres de las columnas, cantidad de registros y tipos de datos del dataframe df_consolidado
df1_consolidado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3960 entries, 0 to 3959
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Fuente  3960 non-null   object 
 1   País    3960 non-null   object 
 2   Año     3960 non-null   int64  
 3   Valor   2663 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 123.9+ KB


Se procede a actuar de acuerdo de acuerdos a los hallazgos clave del EDA en la fase de Extracción y a realizar las transformaciones descritas en la segunda entrega de este proyecto.

## Imputación estructural
se debe realizar la imputación con valor real de 0 para las celdas
nulas con el fin de mantener la relación temporal de las fuentes.

In [7]:
#Remplazo valores nulos de la columna valor
df1_trasform = df1_consolidado.copy()
df1_trasform['Valor'] = df1_trasform['Valor'].fillna(0)
#comprobar si existen valores nulos
print("Nulos restantes:", df1_trasform['Valor'].isna().sum())

Nulos restantes: 0


## Tratamiento de Asimetrías
Aplicación de trasformaciones volumétricas


In [8]:
# convertir de GWh a PTj
df1_trasform['Valor'] =(df1_trasform['Valor']*0.0036).round(6)


In [9]:
print(df1_trasform)

                    Fuente       País   Año       Valor
0                    Hidro  Argentina  2000  103.542563
1     Térmica no renovable  Argentina  2000  165.904690
2                  Nuclear  Argentina  2000   22.237589
3                Geotermia  Argentina  2000    0.000000
4                   Eólica  Argentina  2000    0.125129
...                    ...        ...   ...         ...
3955                Eólica  Venezuela  2021    0.063680
3956                 Solar  Venezuela  2021    0.001312
3957                 Otros  Venezuela  2021    0.000000
3958     Térmica renovable  Venezuela  2021    0.002233
3959                 Total  Venezuela  2021  301.940969

[3960 rows x 4 columns]


# HUB_capacidad_instalada.csv

In [10]:
ruta = "/content/drive/My Drive/Colab Notebooks/Master/1.ETL/entregable3/data/raw/extract_HUB_capacidad_instalada.csv"
df2_consolidado = pd.read_csv(ruta, sep = ';')

In [11]:
# Verificar la cantidad de registros y el número de columnas del dataframe df_consolidado
df2_consolidado.shape

(3960, 4)

Al final se obtienen 3960 registros con 4columnas, la misma información recuperada en la extracción de datos.

In [12]:
# Conocer nombres de las columnas, cantidad de registros y tipos de datos del dataframe df_consolidado
df2_consolidado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3960 entries, 0 to 3959
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Fuente  3960 non-null   object 
 1   País    3960 non-null   object 
 2   Año     3960 non-null   int64  
 3   Valor   2671 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 123.9+ KB


Se procede a actuar de acuerdo de acuerdos a los hallazgos clave del EDA en la fase de Extracción y a realizar las transformaciones descritas en la segunda entrega de este proyecto

## Imputación estructural
se debe realizar la imputación con valor real de 0 para las celdas
nulas con el fin de mantener la relación temporal de las fuentes.

In [13]:
#Remplazo valores nulos de la columna valor
df2_trasform = df2_consolidado.copy()
df2_trasform['Valor'] = df2_trasform['Valor'].fillna(0)
#comprobar si existen valores nulos
print("Nulos restantes:", df2_trasform['Valor'].isna().sum())

Nulos restantes: 0


## Tratamiento de Asimetrías
Aplicación de trasformaciones volumétricas

In [14]:
# convertir de MW a PTj
df2_trasform['Valor'] =(df2_trasform['Valor']*0.031536).round(6)

# HUB_consumo.csv

In [15]:
ruta = "/content/drive/My Drive/Colab Notebooks/Master/1.ETL/entregable3/data/raw/extract_HUB_consumo.csv"
df3_consolidado = pd.read_csv(ruta, sep = ';')

In [16]:
# Verificar la cantidad de registros y el número de columnas del dataframe df_consolidado
df3_consolidado.shape

(2640, 4)

Al final se obtienen 2640 registros con 4columnas, la misma información recuperada en la extracción de datos.

In [17]:
# Conocer nombres de las columnas, cantidad de registros y tipos de datos del dataframe df_consolidado
df3_consolidado.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2640 entries, 0 to 2639
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Sector  2640 non-null   object 
 1   País    2640 non-null   object 
 2   Año     2640 non-null   int64  
 3   Valor   2082 non-null   float64
dtypes: float64(1), int64(1), object(2)
memory usage: 82.6+ KB


Se procede a actuar de acuerdo de acuerdos a los hallazgos clave del EDA en la fase de Extracción y a realizar las transformaciones descritas en la segunda entrega de este proyecto

## Imputación estructural
se debe realizar la imputación con valor real de 0 para las celdas
nulas con el fin de mantener la relación temporal de las fuentes.

In [18]:
#Remplazo valores nulos de la columna valor
df3_trasform = df3_consolidado.copy()
df3_trasform['Valor'] = df2_trasform['Valor'].fillna(0)
#comprobar si existen valores nulos
print("Nulos restantes:", df3_trasform['Valor'].isna().sum())

Nulos restantes: 0


## Tratamiento de Asimetrías
Aplicación de trasformaciones volumétricas

In [19]:
# convertir de GWh a PTj
df3_trasform['Valor'] =(df3_trasform['Valor']*0.0036).round(6)

# Restructuración
Trasformación del formato largo al formato ancho, se debe pivotear
las categorías (fuentes, sector, indicadores) en columnas independientes.

Union de los 3 csv

In [20]:
# Indicador para identificar de donde viene cada variable
df1_trasform['Indicador'] = 'Generacion'
df2_trasform['Indicador'] = 'Capacidad'
df3_trasform['Indicador'] = 'Consumo'

In [21]:
# union 3 csv
df_final = pd.concat([df1_trasform, df2_trasform, df3_trasform], ignore_index=True)
# Unificar fuente y Sector en una sola columna
df_final['Variable'] = df_final['Fuente'].fillna(df_final['Sector'])
# Eliminar  filas "Total" para evitar duplicar datos de los componentes
df_final_sin = df_final[df_final['Variable'] != 'Total'].copy()

In [22]:
#pivot de columnas
df_pivot = df_final_sin.pivot_table(
    index=['País', 'Año'],
    columns=['Indicador', 'Variable'],
    values='Valor',
    aggfunc='sum'
).reset_index()
# aplanar nombre de columnas
df_pivot.columns = ['_'.join(col).strip('_') for col in df_pivot.columns.values]

In [23]:
#comprobar si existen valores nulos
print("Nulos restantes en todo el DataFrame:", df_pivot.isna().sum().sum())

Nulos restantes en todo el DataFrame: 0


Con el fin de estandarizar el dataset para su posterior merge y carga se aplica lo siguiente

In [24]:
#Remover acentos y tildes
df_pivot.columns = df_pivot.columns.str.normalize('NFKD').str.encode('ascii', errors='ignore').str.decode('utf-8')
#Dejar en minusculas
df_pivot.columns = df_pivot.columns.str.lower()
df_pivot.rename(columns={'ano': 'anio'},inplace=True)
df = df_pivot.copy()


# Salida del script
Se guardarán los datos transformados en un csv, para la fase de carga que se trabajará en otro script.

In [26]:
ruta = "/content/drive/My Drive/Colab Notebooks/Master/1.ETL/entregable3/data/transformed/"

# index = False; evita guardar el índice de pandas como una columna adicional
df.to_csv(ruta + "transform_HUB_energia.csv", sep = ";", index = False)
print(df.shape)

(440, 24)
